docs/src/app/(private)/experiments/custom-components.tsx

```diff
@@ -0,0 +1,87 @@
+'use client';
+import * as React from 'react';
+import { useRenderer, RenderProp } from '@base-ui-components/react/use-renderer';
+import styles from './custom-components.module.css';
+
+type Weight = 'light' | 'regular' | 'bold';
+type Size = 'small' | 'medium' | 'large';
+
+type TextState = {
+  weight: Weight;
+  size: Size;
+  excludedProp: boolean;
+};
+
+type TextProps = {
+  className: string | ((state: TextState) => string);
+  weight?: Weight;
+  render?: RenderProp<TextState>;
+  children: React.ReactNode;
+  style?: React.CSSProperties;
+  size?: Size;
+  excludedProp?: boolean;
+};
+
+const Text = React.forwardRef(
+  (props: TextProps, forwardedRef: React.ForwardedRef<HTMLElement>) => {
+    const {
+      className,
+      render,
+      style = {},
+      weight = 'regular',
+      size = 'medium',
+      // Example state prop that we exclude from the style hooks
+      excludedProp = true,
+      ...otherProps
+    } = props;
+
+    const fontWeight = {
+      light: 300,
+      regular: 400,
+      bold: 700,
+    }[weight];
+
+    const state = React.useMemo(
+      () => ({ weight, size, excludedProp }),
+      [weight, size, excludedProp],
+    );
+
+    const { renderElement } = useRenderer({
+      render: render ?? 'p',
+      state,
+      className,
+      ref: forwardedRef,
+      props: {
+        ...otherProps,
+        style: {
+          ...style,
+          fontWeight,
+        },
+      },
+      styleHookMapping: {
+        excludedProp() {
+          return null;
+        },
+      },
```

Comment:
I'm not the biggest fan of this prop name (`styleHookMapping`) - not sure if "style hook" is suitable for a public API, and `Map` instead of `Mapping` might be better

Reply:
The most clear name would be: `stateToDataAttributesMap`, to be very explicit, but it's a strange name 😅 @aarongarciah or @colmtuite do you have some suggestions?

Reply:
Yeah, "hook" has a strong different meaning in React world, so we can think of something else. `stateAttributes`?

Reply:
Technically, you could create any attribute (including `class`), so "dataAttributes" is not entirely correct. We just happen to use only data attributes.

Reply:
I see! `stateAttributes` makes sense then.

Reply:
Renamed to `stateDataAttributes`, I am still missing the mapping in it, but people can figure out how to define them by using the type. Maybe it's even worth adding demo about it on the page. What do you think?